File: PG_Transcript/sol.cpp
#include <bits/stdc++.h>
using namespace std;
struct stu{
    string name;
    int total;
    int C, E, M; // chinese, English, Math
};
bool cmp(const stu &a, const stu &b){
    if(a.total == b.total)
        return a.name < b.name;
    return a.total > b.total;
}
   
int main(){
    int n;
    while(cin >> n){
        vector<stu> all(n);
        for(int i = 0; i < n; i++){
            cin >> all[i].name >> all[i].C >> all[i].E >> all[i].M;
            all[i].total = all[i].C + all[i].E + all[i].M;
        }
        sort(all.begin(), all.end(), cmp);
        int rank = 0, prev = -1, same = 1;
        for(auto &it: all){
            if (it.total == prev)
                same++;
            else {
                rank += same;
                prev = it.total;
                same = 1;
            }
            cout << rank << ' ' << it.name << ' ' << it.total << ' ' << it.C << ' ' << it.E << ' ' << it.M << endl;
        }
        cout << endl;
    }
}

File: PG_Transcript/sol.py
import sys

def main():
    input_lines = sys.stdin.read().strip().split('\n')
    idx = 0
    while idx < len(input_lines):
        if input_lines[idx].strip() == '':
            idx += 1
            continue
        n = int(input_lines[idx].strip())
        idx += 1
        if n == 0:
            break
        students = []
        for _ in range(n):
            line = input_lines[idx].strip().split()
            idx += 1
            name = line[0]
            C, E, M = map(int, line[1:])
            total = C + E + M
            students.append((name, C, E, M, total))
        
        # 排序：總分高到低，若相同依姓名字典序升冪
        students.sort(key=lambda x: (-x[4], x[0]))
        
        rank = 1
        prev = -1
        same = 0
        for i, s in enumerate(students):
            if s[4] == prev:
                same += 1
            else:
                rank += same
                same = 1
                prev = s[4]
            print(rank, s[0], s[4], s[1], s[2], s[3])
        print()

if __name__ == "__main__":
    main()
